File: backend/models/Driver.js
const mongoose = require('mongoose');

const driverSchema = new mongoose.Schema({
  user: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
    unique: true,
    index: true,
  },
  driverUserId: {
    type: Number,
    required: true,
    index: true,
  },
  
  // الربط مع مكتب التاكسي
  office: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'TaxiOffice',
    required: true,
    index: true
  },
  
  carDetails: {
    model: {
      type: String,
      required: true,
      trim: true
    },
    plateNumber: {
      type: String,
      required: true,
      unique: true,
      trim: true,
      uppercase: true
    },
    color: {
      type: String,
      trim: true
    },
    year: {
      type: Number,
      min: 1990,
      max: new Date().getFullYear()
    }
  },
  
  isAvailable: {
    type: Boolean,
    default: true,
    index: true,
  },
  
  rating: {
    type: Number,
    default: 80,
    min: 0,
    max: 100 
  },
  
  numberOfRatings: {
    type: Number,
    default: 0,
    min: 0
  },
  
  profileImageUrl: {
    type: String,
    default: 'https://static.vecteezy.com/system/resources/previews/027/448/973/non_2x/avatar-account-icon-default-social-media-profile-photo-vector.jpg'
  },
  
  earnings: {
    type: Number,
    default: 0,
    min: 0
  },
  
  // معلومات إضافية
  licenseNumber: {
    type: String,
    required: true,
  },
  
  licenseExpiry: {
    type: Date,
    required: true
  },
  
  joinedAt: {
    type: Date,
    default: Date.now
  },
  
  // ✅ إضافة حقل الموقع الحالي للسائق
currentLocation: {
  type: {
    type: String, // لازم يكون "Point"
    enum: ['Point'],
    default: 'Point',
  },
  coordinates: {
    type: [Number],
    default: [35.2137, 31.7683] // ⚠️ مثال: القدس (longitude, latitude)
  }
}
}, { 
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// ✅ إنشاء مؤشر 2dsphere للاستعلامات الجغرافية المكانية
driverSchema.index({ currentLocation: '2dsphere' });

// Middleware لتحديث عدد السائقين عند الحذف
driverSchema.post('save', async function(doc) {
  await updateDriversCount(doc.office);
});

driverSchema.post('remove', async function(doc) {
  await updateDriversCount(doc.office);
});

async function updateDriversCount(officeId) {
  const count = await mongoose.model('Driver').countDocuments({ office: officeId });
  await mongoose.model('TaxiOffice').findByIdAndUpdate(officeId, { driversCount: count });
}

const Driver = mongoose.model('Driver', driverSchema);

module.exports = Driver;

File: lib/models/driver.dart
class Driver {
  final String id; // MongoDB ID
  final int driverUserId;
  final String fullName;
  final String? profileImageUrl;
  final String carModel;
  final String carColor;
  final String carPlateNumber;
  final int? carYear;
  final double rating;
  final int numberOfRatings;
  bool isAvailable;
  final String taxiOfficeId;
  final String phone;
  final String email;
  final double earnings;
  final String licenseNumber;
  final DateTime licenseExpiry;
  final DateTime joinedAt;

  Driver({
    required this.id,
    required this.driverUserId,
    required this.fullName,
    this.profileImageUrl,
    required this.carModel,
    required this.carColor,
    required this.carPlateNumber,
    this.carYear,
    required this.rating,
    required this.numberOfRatings,
    required this.taxiOfficeId,
    required this.phone,
    required this.email,
    required this.earnings,
    required this.isAvailable,
    required this.licenseNumber,
    required this.licenseExpiry,
    required this.joinedAt,
  });

  factory Driver.fromJson(Map<String, dynamic> json) {
    final userDetails = json['user'] as Map<String, dynamic>? ?? {};
    final carDetails = json['carDetails'] as Map<String, dynamic>? ?? {};

    return Driver(
      id: json['_id'] ?? '',
      driverUserId: userDetails['userId'] ?? 0, // ✅ صح
      fullName: userDetails['fullName'] ?? 'Unknown',
      profileImageUrl: json['profileImageUrl'],
      carModel: carDetails['model'] ?? 'N/A',
      carColor: carDetails['color'] ?? 'N/A',
      carPlateNumber: carDetails['plateNumber'] ?? 'N/A',
      carYear: carDetails['year'],
      rating: (json['rating'] as num?)?.toDouble() ?? 0.0,
      numberOfRatings: (json['numberOfRatings'] as num?)?.toInt() ?? 0,
      taxiOfficeId: json['office']?.toString() ?? '',
      isAvailable: json['isAvailable'] ?? false,
      phone: userDetails['phone'] ?? 'N/A',
      email: userDetails['email'] ?? 'N/A',
      earnings: (json['earnings'] as num?)?.toDouble() ?? 0.0,
      licenseNumber: json['licenseNumber'] ?? 'N/A',
      licenseExpiry:
          DateTime.tryParse(json['licenseExpiry'] ?? '') ?? DateTime(2000),
      joinedAt: DateTime.tryParse(json['joinedAt'] ?? '') ?? DateTime(2000),
    );
  }
}
